File: examples/ios/Examples/CustomUserData.m
#import <XCTest/XCTest.h>
#import <Realm/Realm.h>
#import "MyRealmApp.h"

@interface CustomUserData : XCTestCase

@end

@implementation CustomUserData

- (void)testCreateCustomUserData {
    XCTestExpectation *expectation = [self expectationWithDescription:@"it completes"];

    // :code-block-start: create-custom-user-data
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    [app loginWithCredential:[RLMCredentials anonymousCredentials] completion:^(RLMUser *user, NSError *error) {
        if (error != nil) {
            NSLog(@"Failed to log in: %@", error);
            return;
        }
        RLMMongoClient *client = [user mongoClientWithServiceName:@"mongodb-atlas"];
        RLMMongoDatabase *database = [client databaseWithName:@"my_database"];
        RLMMongoCollection *collection = [database collectionWithName:@"users"];
        [collection insertOneDocument:
            @{@"userId": [user identifier], @"favoriteColor": @"pink"}
            completion:^(RLMObjectId *newObjectId, NSError *error) {
                if (error != nil) {
                    NSLog(@"Failed to insert: %@", error);
                    // :hide-start:
                    XCTAssertEqualObjects([error localizedDescription], @"no rule exists for namespace 'my_database.users'");
                    [expectation fulfill];
                    // :hide-end:
                }
                NSLog(@"Inserted custom user data document with object ID: %@", newObjectId);
        }];
    }];
    // :code-block-end:

    [self waitForExpectationsWithTimeout:10 handler:^(NSError *error) {
        NSLog(@"Expectation failed: %@", error);
    }];
}

- (void)testReadCustomUserData {
    XCTestExpectation *expectation = [self expectationWithDescription:@"it completes"];

    // :code-block-start: read-custom-user-data
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    [app loginWithCredential:[RLMCredentials anonymousCredentials] completion:^(RLMUser *user, NSError *error) {
        if (error != nil) {
            NSLog(@"Failed to log in: %@", error);
            return;
        }
        
        // If the user data has been refreshed recently, you can access the
        // custom user data directly on the user object
        NSLog(@"User custom data: %@", [user customData]);
        
        // Refresh the custom data
        [user refreshCustomDataWithCompletion:^(NSDictionary *customData, NSError *error) {
            if (error != nil) {
                NSLog(@"Failed to refresh custom user data: %@", error);
                return;
            }
            NSLog(@"Favorite color: %@", customData[@"favoriteColor"]);
            // :hide-start:
            [expectation fulfill];
            // :hide-end:
        }];
    }];
    // :code-block-end:

    [self waitForExpectationsWithTimeout:10 handler:^(NSError *error) {
        NSLog(@"Expectation failed: %@", error);
    }];
}

- (void)testUpdateCustomUserData {
    XCTestExpectation *expectation = [self expectationWithDescription:@"it completes"];

    // :code-block-start: update-custom-user-data
    RLMApp *app = [RLMApp appWithId:YOUR_REALM_APP_ID];
    [app loginWithCredential:[RLMCredentials anonymousCredentials] completion:^(RLMUser *user, NSError *error) {
        if (error != nil) {
            NSLog(@"Failed to log in: %@", error);
            return;
        }
        RLMMongoClient *client = [user mongoClientWithServiceName:@"mongodb-atlas"];
        RLMMongoDatabase *database = [client databaseWithName:@"my_database"];
        RLMMongoCollection *collection = [database collectionWithName:@"users"];

        // Update the user's custom data document
        [collection updateOneDocumentWhere:@{@"userId": [user identifier]}
            updateDocument: @{@"favoriteColor": @"cerulean"}
            completion:^(RLMUpdateResult *updateResult, NSError *error) { 
                if (error != nil) {
                    NSLog(@"Failed to insert: %@", error);
                    // :hide-start:
                    XCTAssertEqualObjects([error localizedDescription], @"no rule exists for namespace 'my_database.users'");
                    [expectation fulfill];
                    // :hide-end:
                }
                NSLog(@"Matched: %lu, modified: %lu", [updateResult matchedCount], [updateResult modifiedCount]); 
        }];
    }];
    // :code-block-end:

    [self waitForExpectationsWithTimeout:10 handler:^(NSError *error) {
        NSLog(@"Expectation failed: %@", error);
    }];
}

@end

File: examples/ios/Examples/CustomUserData.swift
import XCTest
import RealmSwift

class CustomUserData: XCTestCase {

    func testCreateCustomUserData() {
        let expectation = XCTestExpectation(description: "it completes")

        // :code-block-start: create-custom-user-data
        let appId = YOUR_REALM_APP_ID // replace this with your App ID
        let app = App(id: appId)
        app.login(credentials: Credentials.anonymous) { (result) in
            switch result {
            case .failure(let error):
                print("Failed to log in: \(error.localizedDescription)")
            case .success(let user):
                let client = user.mongoClient("mongodb-atlas")
                let database = client.database(named: "my_database")
                let collection = database.collection(withName: "users")

                // Insert the custom user data object
                collection.insertOne([
                    "userId": AnyBSON(user.id),
                    "favoriteColor": "pink"
                    ]) { (result) in
                    switch result {
                    case .failure(let error):
                        print("Failed to insert document: \(error.localizedDescription)")
                        // :hide-start:
                        XCTAssertEqual(error.localizedDescription, "no rule exists for namespace 'my_database.users'")
                        expectation.fulfill()
                        // :hide-end:
                    case .success(let newObjectId):
                        print("Inserted custom user data document with object ID: \(newObjectId)")
                    }
                }
            }
        }
        // :code-block-end:
        wait(for: [expectation], timeout: 10)
    }

    func testReadCustomUserData() {
        let expectation = XCTestExpectation(description: "it completes")

        // :code-block-start: read-custom-user-data
        let appId = YOUR_REALM_APP_ID // replace this with your App ID
        let app = App(id: appId)
        app.login(credentials: Credentials.anonymous) { (result) in
            switch result {
            case .failure(let error):
                print("Failed to log in: \(error.localizedDescription)")
            case .success(let user):
                // If the user data has been refreshed recently, you can access the
                // custom user data directly on the user object
                print("User custom data: \(user.customData)")

                // Refresh the custom user data
                user.refreshCustomData() { (result) in
                    switch result {
                    case .failure(let error):
                        print("Failed to refresh custom data: \(error.localizedDescription)")
                    case .success(let customData):
                        // favoriteColor was set on the custom data.
                        print("Favorite color: \(customData["favoriteColor"] ?? "not set")")
                        // :hide-start:
                        expectation.fulfill()
                        // :hide-end:
                        return
                    }
                }
            }
        }
        // :code-block-end:
        wait(for: [expectation], timeout: 10)
    }

    func testUpdateCustomUserData() {
        let expectation = XCTestExpectation(description: "it completes")

        // :code-block-start: update-custom-user-data
        let appId = YOUR_REALM_APP_ID // replace this with your App ID
        let app = App(id: appId)
        app.login(credentials: Credentials.anonymous) { (result) in
            switch result {
            case .failure(let error):
                print("Failed to log in: \(error.localizedDescription)")
            case .success(let user):
                // Access the custom user document remotely to update it.
                let client = user.mongoClient("mongodb-atlas")
                let database = client.database(named: "my_database")
                let collection = database.collection(withName: "users")
                collection.updateOneDocument(
                    filter: ["userId": AnyBSON(user.id)],
                    update: ["favoriteColor": "cerulean"]
                ) { (result) in
                    switch result {
                    case .failure(let error):
                        print("Failed to update: \(error.localizedDescription)")
                        // :hide-start:
                        XCTAssertEqual(error.localizedDescription, "no rule exists for namespace 'my_database.users'")
                        expectation.fulfill()
                        // :hide-end:
                        return
                    case .success(let updateResult):
                        // User document updated.
                        print("Matched: \(updateResult.matchedCount), updated: \(updateResult.modifiedCount)")
                    }
                }
            }
        }
        // :code-block-end:
        wait(for: [expectation], timeout: 10)
    }

}
